swiss_tournament/step/round_generator.py

```diff
@@ -0,0 +1,98 @@
+from typing import Dict, List
+
+import networkx as nx
+
+from swiss_tournament.data.player import BYE, Player
+from swiss_tournament.data.result import Result
+from swiss_tournament.data.tournament import Tournament
+from swiss_tournament.step.helper_mappings import player_points_map, player_color_map
+
+NEG_INF = -100000000
+
+
+class RoundGenerator:
+
+    @staticmethod
+    def pairing_cost(
+            a: Player,
+            b: Player,
+            player_scores: Dict[Player, float],
+            player_colors: Dict[Player, tuple[int, int]]
+    ) -> float:
+        const = 100
+        a_score = player_scores.get(a) or 0
+        b_score = player_scores.get(b) or 0
+        a_whites, a_blacks = player_colors.get(a) or (0, 0)
+        b_whites, b_blacks = player_colors.get(b) or (0, 0)
+        score_diff = abs(a_score - b_score)
+        color_diff = abs(a_whites - b_blacks) + abs(a_blacks - b_whites)
+        if a == BYE:
+            return - (const * 20 * b_score)
+        elif b == BYE:
+            return - (const * 20 * a_score)
+        else:
+            return -(const * score_diff + color_diff)
+
+    def _populate_graph(
+            self,
+            tournament: Tournament,
+            player_scores: Dict[Player, float],
+            player_colors: Dict[Player, tuple[int, int]]
+    ) -> nx.Graph:
+        graph = nx.Graph()
+        graph.clear()
+
+        players = tournament.players
+        # One BYE in case we don't have even players
+        if len(players) % 2 != 0:
+            players.append(BYE)
+
+        # One node per player
+        for player in players:
+            graph.add_node(player)
+
+        # One edge for each possible match
+        for a in players:
+            for b in players:
+                if a != b:
+                    graph.add_edge(a, b, weight=RoundGenerator.pairing_cost(a, b, player_scores, player_colors))
+
+        # Eliminate the option to repeat pairings
+        for played_round in tournament.rounds:
+            for result in played_round.pairing:
+                graph.remove_edge(result.white, result.black)
```

Comment:
I do not think this is strictly necessary. It can happen that in a setting with a certain number of rounds, some players may need to play each other again. I am thinking of a bullet tournament where the tournament ends after a specific time is up, like the standard open tournaments on Lichess. Instead of removing this edge from the graph, I would assign a high cost to it (like in the original implementation). This will avoid pairing the same players again twice, but it would still have a non-zero probability.

Reply:
ok, I will do that. I added this impediment since in swiss, pairing again with the same person isn't allowed. But we can still assign a really high value to allow more rounds if needed